File: src/tonemap/RevisedTumblinRushmeierToneMap.cpp
#include "java/lang/Math.h"
#include "tonemap/RevisedTumblinRushmeierToneMap.h"

/**
References:

[TUMB1999b] J. Tumblin, J.K. Hodgins, B.K. Guenter. Two Methods for Display of High
Contrast Images, ACM Transactions on Graphics, 18:1, 1999, pp. 56-94.
*/

static float globalG;
static float globalComp;
static float globalDisp;
static float globalLwa;
static float globalLdaTumb;

RevisedTumblinRushmeierToneMap::RevisedTumblinRushmeierToneMap() {
}

RevisedTumblinRushmeierToneMap::~RevisedTumblinRushmeierToneMap() {
}

void
RevisedTumblinRushmeierToneMap::init() {
    float lwa = GLOBAL_toneMap_options.realWorldAdaptionLuminance;
    float ldmax = GLOBAL_toneMap_options.maximumDisplayLuminance;
    float cmax = GLOBAL_toneMap_options.maximumDisplayContrast;
    globalLdaTumb = ldmax / java::Math::sqrt(cmax);

    globalG = stevensGamma(lwa) / stevensGamma(globalLdaTumb);
    float gwd = stevensGamma(lwa) / (1.855f + 0.4f * java::Math::log(globalLdaTumb));
    globalComp = java::Math::pow(java::Math::sqrt(cmax), gwd - 1) * globalLdaTumb;
    globalDisp = globalComp / ldmax;
}

float
RevisedTumblinRushmeierToneMap::stevensGamma(float lum) {
    if ( lum > 100.0 ) {
        return 2.655f;
    } else {
        return 1.855f + 0.4f * java::Math::log10(lum + 2.3e-5f);
    }
}

static ColorRgb
revisedTRScaleForComputations(ColorRgb radiance) {
    float rwl;
    float scale;

    rwl = radiance.luminance();

    if ( rwl > 0.0 ) {
        scale = globalComp * java::Math::pow(rwl / globalLwa, globalG) / rwl;
    } else {
        scale = 0.0;
    }

    radiance.scale(scale);
    return radiance;
}

static ColorRgb
revisedTRScaleForDisplay(ColorRgb radiance) {
    float rwl = (float)M_PI * radiance.luminance();
    float eff = getLuminousEfficacy();
    radiance.scale(eff * (float)M_PI);

    float scale;
    if ( rwl > 0.0 ) {
        scale = globalDisp * java::Math::pow(rwl / globalLwa, globalG) / rwl;
    } else {
        scale = 0.0f;
    }

    radiance.scale(scale);
    return radiance;
}

OldToneMap GLOBAL_toneMap_revisedTumblinRushmeier = {
    "Revised Tumblin/Rushmeier's Mapping",
    "RevisedTR",
    3,
    nullptr,
    revisedTRScaleForComputations,
    revisedTRScaleForDisplay
};

File: typescript/src/main/vsdk/toolkit/tonemap/RevisedTumblinRushmeierToneMap.ts
import { Cie } from "../common/Cie";
import { ColorRgb } from "../common/ColorRgb";
import { ToneMap } from "./ToneMap";
import { ToneMappingContext } from "./ToneMappingContext";

export class RevisedTumblinRushmeierToneMap extends ToneMap {
  private g: number;
  private comp: number;
  private display: number;
  private lwaRTR: number;
  private ldaRTR: number;

  private static stevensGamma(lum: number): number {
    if (lum > 100.0) {
      return 2.655;
    }
    return 1.855 + 0.4 * globalThis.Math.log10(lum + 2.3e-5);
  }

  public constructor() {
    super();
    this.g = 0.0;
    this.comp = 0.0;
    this.display = 0.0;
    this.lwaRTR = 0.0;
    this.ldaRTR = 0.0;
  }

  public override init(toneMapOptions: ToneMappingContext): void {
    const lwa = toneMapOptions.realWorldAdaptionLuminance;
    const maximumDisplayLuminance = toneMapOptions.maximumDisplayLuminance;
    const maximumDisplayContrast = toneMapOptions.maximumDisplayContrast;
    this.ldaRTR = maximumDisplayLuminance / globalThis.Math.sqrt(maximumDisplayContrast);

    this.g = RevisedTumblinRushmeierToneMap.stevensGamma(lwa) / RevisedTumblinRushmeierToneMap.stevensGamma(this.ldaRTR);
    const gwd = RevisedTumblinRushmeierToneMap.stevensGamma(lwa) / (1.855 + 0.4 * globalThis.Math.log10(this.ldaRTR));
    this.comp = globalThis.Math.pow(globalThis.Math.sqrt(maximumDisplayContrast), gwd - 1) * this.ldaRTR;
    this.display = this.comp / maximumDisplayLuminance;
  }

  public override scaleForComputations(radiance: ColorRgb): ColorRgb {
    const rwl = radiance.luminance();
    let scale: number;

    if (rwl > 0.0) {
      scale = this.comp * globalThis.Math.pow(rwl / this.lwaRTR, this.g) / rwl;
    }
    else {
      scale = 0.0;
    }

    radiance.scale(scale);
    return radiance;
  }

  public override scaleForDisplay(radiance: ColorRgb): ColorRgb {
    const rwl = globalThis.Math.PI * radiance.luminance();
    const eff = Cie.getLuminousEfficacy();
    radiance.scale(eff * globalThis.Math.PI);

    let scale: number;
    if (rwl > 0.0) {
      scale = this.display * globalThis.Math.pow(rwl / this.lwaRTR, this.g) / rwl;
    }
    else {
      scale = 0.0;
    }

    radiance.scale(scale);
    return radiance;
  }
}
